src/Agent.Listener/JobDispatcher.cs

```diff
@@ -1,44 +1,231 @@
 ﻿using Microsoft.TeamFoundation.DistributedTask.WebApi;
 using Microsoft.VisualStudio.Services.Agent.Util;
 using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.IO;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.VisualStudio.Services.Agent.Listener
 {
     [ServiceLocator(Default = typeof(JobDispatcher))]
-    public interface IJobDispatcher : IDisposable, IAgentService
+    public interface IJobDispatcher : IAgentService
     {
-        Task<int> RunAsync(JobRequestMessage message, CancellationToken jobRequestCancellationToken);
+        void Run(JobRequestMessage message);
+        bool Cancel(JobCancelMessage message);
+        Task WaitAsync(CancellationToken token);
+        Task Shutdown();
     }
 
     public sealed class JobDispatcher : AgentService, IJobDispatcher
     {
+        private class Dispatch : IDisposable
+        {
+            public Dispatch(Guid jobId, long requestId)
+            {
+                JobId = jobId;
+                RequestId = requestId;
+                WorkerCancellationTokenSource = new CancellationTokenSource();
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+
+            private void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    if (WorkerCancellationTokenSource != null)
+                    {
+                        WorkerCancellationTokenSource.Cancel();
+                        WorkerCancellationTokenSource = null;
+                    }
+                }
+            }
+
+            public long RequestId { get; }
+            public Guid JobId { get; }
+            public Task WorkerDispatch { get; set; }
+            public CancellationTokenSource WorkerCancellationTokenSource { get; private set; }
+        }
+
         private static readonly string _workerProcessName = $"Agent.Worker{IOUtil.ExeExtension}";
 
+        private int _poolId;
+
+        private readonly Queue<Dispatch> _workerDispatchQueue = new Queue<Dispatch>();
```

Comment:
You usages of queue seem to have the potential to be used in a non-thread-safe manner.